### PR TITLE
Cumulative sum functionality

### PR DESCRIPTION
**Pull request by hjonasson**
Issue from Sept. 2019 [here](https://github.com/josdejong/mathjs/issues/1633#issue-498532285) for a cumsum function.

It handles inputs in the same way as the math.js `sum` function and n-dimensional cumsum in the same way as numpy's implementation.

### CODE REVIEW

File: src/utils/switch.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Transpose a matrix
+ * @param {Array} mat
+ * @returns {Array} ret
+ * @private
+ */
+export function _switch (mat) {
```
Comment (josdejong) [inline src/utils/switch.js]: Makes a lot of sense de-duplicating this helper function :smile:

File: test/unit-tests/function/statistics/cumsum.test.js [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+      [3, 0, 5],
+      [-1, 11, 9]
+    ]), [
+      [1, 4, 7],
+      [4, 4, 12],
+      [3, 15, 21]
+    ])
+  })
+
+  //   TODO: make a dense matrix return a dense matrix, I had issues depending on it in cumsum.js
```
Comment (josdejong) [inline test/unit-tests/function/statistics/cumsum.test.js]: Is this TODO still relevant?

**Comment by josdejong** [comment]
Thanks for implementing `cumsum` @hjonasson, looks good :+1: ! I like it that it that the API is the same as `sum`.

A few remarks:

1. There is one unit test failing because the library expects to have embedded docs for each of the functions. This is a JSON object in a JS file in `src/expression/embeddedDocs/function`, you can take [`sum.js`](https://github.com/josdejong/mathjs/blob/develop/src/expression/embeddedDocs/function/statistics/sum.js) as example, shouldn't be too hard.
2. It looks like there is support for passing along _which_ dimension you would like to sum values, `function _ncumSumDim (array, dim)`, but I don't see this reflected in the docs or unit tests, and when trying it out it returns an empty array. Is this supposed to work?

File: src/utils/switch.js (hunk 1 above)
Comment (hjonasson) [reply src/utils/switch.js]: I think I introduced the re-use of this function and so felt responsible to not duplicate it.

File: test/unit-tests/function/statistics/cumsum.test.js (hunk 2 above)
Comment (hjonasson) [reply test/unit-tests/function/statistics/cumsum.test.js]: Nope. Removed. Good catch.

File: test/unit-tests/function/statistics/cumsum.test.js [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        [150, 310],
+        [170, 350]
+      ],
+      [
+        [190, 390],
+        [210, 430],
+        [230, 470]
+      ]
+    ])
+  })
```
Comment (hjonasson) [inline test/unit-tests/function/statistics/cumsum.test.js]: @josdejong here are tests for passing in dimensions

**Comment by hjonasson** [comment]
@josdejong I can't for the life of me comment on your comment. Thanks for the help. 

I added embedded docs, with a comment/question. Please resolve if it is a non-issue. 

I believe I have tests for taking in dimensions, I tagged you in it, let me know if I am missing something there.

File: src/expression/embeddedDocs/function/statistics/cumsum.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  category: 'Statistics',
+  syntax: [
+    'cumsum(a, b, c, ...)',
+    'cumsum(A)'
+  ],
+  description: 'Compute the sum of all values.',
+  examples: [
+    'cumsum(2, 3, 4, 1)',
+    'cumsum([2, 3, 4, 1])',
+    'cumsum([2, 5; 4, 3])'
```
Comment (hjonasson) [inline src/expression/embeddedDocs/function/statistics/cumsum.js]: I got this from sum.js in the same folder. Should the semicolon be there?

Comment (hjonasson) [reply src/expression/embeddedDocs/function/statistics/cumsum.js]: @josdejong

File: src/expression/embeddedDocs/function/statistics/cumsum.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  ],
+  seealso: [
+    'max',
```
Comment (hjonasson) [inline src/expression/embeddedDocs/function/statistics/cumsum.js]: Should I add a `'cumsum'` to the `seealso`s in other files here? @josdejong

**Comment by m93a** [comment]
Please let's not name it cumsum.

**Comment by hjonasson** [comment]
It is a widely used name for cumulative sum, not only in programming. Some
people do giggle the first time they see it if that is the problem.

Þann sun., 15. mar. 2020, 16:44 Michal Grňo skrifaði <
notifications@github.com>:

**Comment by hjonasson** [comment]
@josdejong Is there anything further I can do with this PR?

File: src/expression/embeddedDocs/function/statistics/cumsum.js (hunk 5 above)
Comment (josdejong) [reply src/expression/embeddedDocs/function/statistics/cumsum.js]: Yes, good idea, would be nice to add a reference for example in `sum`.

File: src/expression/embeddedDocs/function/statistics/cumsum.js (hunk 4 above)
Comment (josdejong) [reply src/expression/embeddedDocs/function/statistics/cumsum.js]: The semicolon makes it a 2x2 matrix.

**Comment by josdejong** [comment]
@hjonasson sorry for the late reply. We're adapting here (Netherlands) to restrictions against Corona, I put my open source projects are a bit on the backburner.

I did overlook the tests you've added to test dimensions, that works fine. I stumbled across an issue in this regard:

```js
// works:
math.cumsum([[2,6],[4,10]], 0) // [ [ 2, 6 ], [ 6, 16 ] ]
math.cumsum([[2,6],[4,10]], 0) // [ [ 2, 6 ], [ 6, 16 ] ]
math.cumsum([[2,6],[4,10]], 1) // [ [ 2, 8 ], [ 4, 14 ] ]

// does not work:
math.evaluate('cumsum([2,6; 4,10], 1)') // [ ]
```

I'm not sure what can cause this, it looks odd. Functions working with dimensions like `sum` and `cumsum` need a _transform_ since the expression parser is one-based whilst the functions themselves are zero-based (long story, there are some discussions about it in this issues section, like #23). You can have a look and see if you can create a `cumsum.transform.js` similar to `sum.transform.js`. But I can have a look too, this transform-thing can be intimidating, and I know probably better what to look for exactly.

**Comment by hjonasson** [comment]
Late replies here as well:D I have looked into the expression parsing of mathjs now. As you point out, this code doesn't fully support that yet. Cumsum across dimensions via expressions is missing in the PR, I will see if I can figure it out.

**Comment by josdejong** [comment]
Cool, thanks for the update. Let me know if you need help.

**Comment by josdejong** [comment]
ping @hjonasson did you have any time already to look at the last thingies to finish this PR?

**Comment by hjonasson** [comment]
I gave it one shot before becoming too busy with corona related things
myself. I tried to understand why the same cases as code and as strings
didn't produce the same results. Perhaps you could guide me to something
about the parsing? Or is it obvious what is going wrong here?

Þann sun., 3. maí 2020, 10:42 Jos de Jong skrifaði <notifications@github.com

**Comment by josdejong** [comment]
> I tried to understand why the same cases as code and as strings
didn't produce the same results. Perhaps you could guide me to something
about the parsing? Or is it obvious what is going wrong here?

If you mean the difference between the plain JavaScript API and the expression parser: there are a number of differences there, the expression parser is NOT JavaScript. There is some documentation about ithere: https://mathjs.org/docs/expressions/syntax.html#differences-from-javascript. 

Again, it's perfectly fine for me to work out creating a transform function for `cumsum`, just let me know.

**Comment by hjonasson** [comment]
Perhaps you could look at that then? I believe there is some polishing I need to do for the documentation still.

**Comment by josdejong** [comment]
👍 will do.

**Comment by josdejong** [comment]
@hjonasson I've had a look at your work, and added some more tests and did some fixes, see:

commit: https://github.com/josdejong/mathjs/commit/890b90b7119c6e7f707def39458a42dd4ba8adec
branch: https://github.com/josdejong/mathjs/tree/hjonasson-develop 

It is ready to be merged, but there is one small piece of code that gives me a hard time understanding whats going on:

```js
return array.map((sum => value => sum = add(sum, value))(_initialCumsumValue(array[0]))) // eslint-disable-line
```
Can you rewrite this for me please?

**Comment by hjonasson** [comment]
Will do

Þann mið., 27. maí 2020, 10:35 Jos de Jong skrifaði <
notifications@github.com>:

**Comment by josdejong** [comment]
Thanks! It may be easiest to simply paste the code here, then I can add it to my version of your PR, else we get two diverting paths.

**Comment by hjonasson** [comment]
I have somewhat different variable names from you it seems but here is a
bit of a refactoring that I hope makes the code clearer.

function _cumulativeSummation (sum) {
return value => sum = add(sum, value) // eslint-disable-line
}

function _cumsummap (array) {
const arrayIsMatrix = isMatrix(array)
const tempArray = arrayIsMatrix ? array.valueOf() : array
const initialValue = _initialCumsumValue(tempArray[0])
const cumsumArray = tempArray.map(_cumulativeSummation(initialValue))
return isMatrix(array) ? new DenseMatrix(cumsumArray) : cumsumArray
}

**Comment by hjonasson** [comment]
```JS
  function _cumulativeSummation (sum) {
    return value => sum = add(sum, value) // eslint-disable-line
  }

  function _cumsummap (array) {
    const arrayIsMatrix = isMatrix(array)
    const tempArray = arrayIsMatrix ? array.valueOf() : array
    const initialValue = _initialCumsumValue(tempArray[0])
    const cumsumArray = tempArray.map(_cumulativeSummation(initialValue))
    return isMatrix(array) ? new DenseMatrix(cumsumArray) : cumsumArray
  }
```

**Comment by josdejong** [comment]
Thanks Hjörtur! Now I understand. It's very nifty, compact code, but I prefer having easy to understand code instead :). I've rewritten the function now, and created a PR: https://github.com/josdejong/mathjs/pull/1870

I've indeed rewritten some parts of the code, for example converting Matrix into Array and back is done in an earlier stage, so that most of the code simply only has to deal with Array.

Can you have a look if all is ok this way?

There is one thing that I'm in doubt about (didn't dive into it yet): I would expect `add(sum, value)` in `_cumsummap` to only sum scalar values, but it looks like (sometimes?) `value` is an Array: if you replace `add` with `addScalar`, some tests fail. Any idea?

**Comment by josdejong** [comment]
Ping @hjonasson. Would be great if you can comment on my previous comment: https://github.com/josdejong/mathjs/pull/1765#issuecomment-634892499

**Comment by hjonasson** [comment]
Sounds like it might be matrices. Let me pull your branch and see what is breaking.

**Comment by josdejong** [comment]
@hjonasson I see this PR is still open. It was almost finished, would be great to get it merged :). Any chance to look into the last bits and pieces?

**Comment by gwhitney** [comment]
As far as I could see, the only remaining question was why it was necessary to use add and not addScalar internally. This is because in the case of cumsumming, say, a 2d matrix along the first dimension, the rows are literally added, i.e.
`cumsum([1,3; 5,7])` is literally produced by letting the first row be `[1,3]` and the second row be `add([1,3],[5,7])` which produces the correct result `[1,3; 6,10]`.

So I rebased on develop, added a TypeScript declaration, tweaked the documentation just slightly, and also avoided the potentially slightly confusing subtract-and-then-add step in `_cumsummap`.

Will push to #1870 and then merge.

**Comment by gwhitney** [comment]
Closing this in favor of #1870 that was just merged.

**Comment by josdejong** [comment]
👍  thanks Glen

**Pull request closed without merging**